spyderlib/utils/introspection/plugin_server.py

```diff
@@ -4,101 +4,95 @@
+    for interacting with an object.
     """
 
-    def __init__(self, client_port, plugin_name):
-        mod_name = plugin_name + '_plugin'
-        mod = __import__('spyderlib.utils.introspection.' + mod_name,
-                         fromlist=[mod_name])
-        cls = getattr(mod, '%sPlugin' % plugin_name.capitalize())
-        plugin = cls()
-        plugin.load_plugin()
-        self.tlast = time.time()
-        self.plugin = plugin
-
-        self._client_port = int(client_port)
-        sock, self.server_port = connect_to_port()
-        sock.listen(2)
-        atexit.register(sock.close)
-        self._server_sock = sock
-
-        self.queue = Queue.Queue()
-        self._listener = threading.Thread(target=self.listen)
-        self._listener.setDaemon(True)
-        self._listener.start()
-
-        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        sock.connect(("127.0.0.1", self._client_port))
-        write_packet(sock, self.server_port)
-        sock.close()
-
-    def listen(self):
-        """Listen for requests"""
-        while True:
-            try:
-                conn, _addr = self._server_sock.accept()
-            except socket.error as e:
-                badfd = errno.WSAEBADF if os.name == 'nt' else errno.EBADF
-                extra = errno.WSAENOTSOCK if os.name == 'nt' else badfd
-                if e.args[0] in [errno.ECONNABORTED, badfd, extra]:
-                    return
-                # See Issue 1275 for details on why errno EINTR is
-                # silently ignored here.
-                eintr = errno.WSAEINTR if os.name == 'nt' else errno.EINTR
-                if e.args[0] == eintr:
-                    continue
-                raise
-            self.queue.put(read_packet(conn))
+    def __init__(self, port, *args):
+        self.port = port
+        self.object = self.initialize(*args)
+        self.context = zmq.Context()
+        self.socket = self.context.socket(zmq.PAIR)
+        self.socket.connect("tcp://localhost:%s" % port)
+        self.socket.send_pyobj(port)
+
+    def initialize(self, plugin_name):
+        """Initialize the object and return it.
+        """
+        return object()
 
     def run(self):
-        """Handle requests"""
+        """Handle requests from the client.
+        """
         while 1:
-            # Get most recent request
-            request = None
+            # Poll for events, handling a timeout.
+            events = self.socket.poll(TIMEOUT)
+            if events == 0:
+                print('Timed out')
+                return
+            # Drain all exising requests, handling quit and heartbeat.
+            requests = []
             while 1:
-                try:
-                    request = self.queue.get(True, 0.01)
-                except Queue.Empty:
+                request = self.socket.recv_pyobj()
+                if request['func_name'] == 'server_quit':
+                    print('Quitting')
```

Comment:
This is being run in a separate process, with a separate stdout..  We want the print so we can capture it from the client in the event of an error.